File: lib/configparser/benches/bench.rs
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::io::Write;

use bytes::Bytes;
use minibench::{bench, elapsed};

use configparser::config::ConfigSet;

fn main() {
    bench("parse 645KB file", || {
        let mut config_file = Vec::new();
        for _ in 0..100 {
            for section in b'a'..b'z' {
                config_file
                    .write(format!("[{ch}{ch}{ch}{ch}]\n", ch = section as char).as_bytes())
                    .unwrap();
                for name in b'a'..b'z' {
                    config_file
                        .write(
                            format!("{ch}{ch}{ch} = {ch}{ch}{ch}\n", ch = name as char).as_bytes(),
                        )
                        .unwrap();
                }
            }
        }
        elapsed(|| {
            let mut cfg = ConfigSet::new();
            cfg.parse(Bytes::from(&config_file[..]), &"bench".into());
        })
    });
}

File: eden/fs/benchharness/Bench.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#include "eden/fs/benchharness/Bench.h"
#include <inttypes.h>
#include <stdio.h>
#include <time.h>

namespace facebook {
namespace eden {

uint64_t getTime() noexcept {
  timespec ts;
  // CLOCK_MONOTONIC is subject in NTP adjustments. CLOCK_MONOTONIC_RAW would be
  // better but these benchmarks are short and reading CLOCK_MONOTONIC takes 20
  // ns and CLOCK_MONOTONIC_RAW takes 130 ns.
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return ts.tv_sec * 1000000000 + ts.tv_nsec;
}

StatAccumulator measureClockOverhead() noexcept {
  constexpr int N = 10000;

  StatAccumulator accum;

  uint64_t last = getTime();
  for (int i = 0; i < N; ++i) {
    uint64_t next = getTime();
    uint64_t elapsed = next - last;
    accum.add(elapsed);
    last = next;
  }

  return accum;
}

} // namespace eden
} // namespace facebook
